src/Confluent.Kafka.Avro/GenericSerializerImpl.cs

```diff
@@ -111,18 +111,20 @@ public byte[] Serialize(string topic, GenericRecord data)
                 var subjectSchemaPair = new KeyValuePair<string, string>(subject, writerSchemaString);
                 if (!registeredSchemas.Contains(subjectSchemaPair))
                 {
+                    int newSchemaId;
                     // first usage: register/get schema to check compatibility
                     if (autoRegisterSchema)
                     {
-                        schemaIds.Add(
-                            writerSchemaString,
-                            schemaRegistryClient.RegisterSchemaAsync(subject, writerSchemaString).ConfigureAwait(false).GetAwaiter().GetResult());
+                        newSchemaId = schemaRegistryClient.RegisterSchemaAsync(subject, writerSchemaString).ConfigureAwait(false).GetAwaiter().GetResult();
                     }
                     else
                     {
-                        schemaIds.Add(
-                            writerSchemaString,
-                            schemaRegistryClient.GetSchemaIdAsync(subject, writerSchemaString).ConfigureAwait(false).GetAwaiter().GetResult());
+                        newSchemaId = schemaRegistryClient.GetSchemaIdAsync(subject, writerSchemaString).ConfigureAwait(false).GetAwaiter().GetResult();
+                    }
+
+                    if (!schemaIds.ContainsKey(writerSchemaString))
+                    {
+                        schemaIds.Add(writerSchemaString, newSchemaId);
                     }
```

Comment:
I also thought about that but wasn't sure that's why i left it open. This can happen only if schema is deleted in schema registry and is inserted again and the producer wants to send data to new topic that is not in local subjects cache. It will fail 100 times before that (maybe in the consumers), but if we already have the new id , why don't we just update it and it will continue working.

Reply:
nice - I didn't think through that scenario: a schema is deleted, then re-registered, then the client tries to produce to a topic it hasn't yet produced to. In light of that, I believe we want the additional check I propose above, but before throwing the exception, we clear the local schema cache. that way, in the above scenario, we get one message produce failure (which i think is ok, good even, in this scenario), and subsequent produce requests work as expected.